### DISCUSSION TITLE
Error validating datasource `db`: You must provide a nonempty URL. How to avoid prisma looking up datasource URL on schema.prisma?

### DISCUSSION THREAD

**Discussion by TheVoid-0** [prisma/prisma#11994] [Q&A] [answered]
Hello, I will try to be fast and clear. I have a multi-tenant application, wich means every prismaClient I instantiate has a different datasource url passed to it via constructor, apparently it's working fine, but every now an then (and this is VERY inconsistent) I get an error saying `invalid prisma.<model>.findMany invocation: error: Error validating datasource 'db': You must provide a nonempty URL` I indeed do not have a URL setted in datasource, because that would be impossible, since I only know what DB I'm using on the runtime.

Like I said, this is very inconsistent, sometimes the api runs for hours before the first error happens (in the same code location) and sometimes it's the first thing that happens when the api starts, but it seems to occurs earlier on linux environments for some reason (testing with Docker). I need a way to avoid prisma trying to load a useless URL since the correct one is already being provide upon the client construction, what is the need to looking up on the schema?

Here is the complete error log:

```
Invalid `prisma.<model>.findFirst()` invocation:


  error: Error validating datasource `db`: You must provide a nonempty URL
  -->  schema.prisma:8
   |
 7 |   provider = "mysql"
 8 |   url      = ""
   |

Validation Error Count: 1
    at Object.request (C:\development\api\node_modules\@prisma\client\runtime\index.js:39074:15)
    at TenantConnection._request (C:\development\api\node_modules\@prisma\client\runtime\index.js:40859:18)
    at BudgetService.get (C:\development\api\src\budgets\services\budgets.service.ts:60:28)
    at C:\development\api\node_modules\@nestjs\core\router\router-execution-context.js:46:28
    at C:\development\api\node_modules\@nestjs\core\router\router-proxy.js:9:17

```
Thanks in advance.

**Comment by tarazena** [comment]
why not set it to dummy url? as long as you pass the url in every client creation i think it should be fine.

**Reply by TheVoid-0** [reply]
I would use this in case there was no better solution, I don't have any guarantees that my "test" database will exists forever, since a tenant can drop his database at any moment I would need to keep a dummy database up just for this. And anyway, why does prisma need to verify the datasource url if it was already provided on constructor call? And I also do not know what he will do with this url, will prisma connect to this dummy db?

Thanks for your answer by the way.

**Reply by janpio** [reply]
Just provide one that does not work, but fulfills the requirements of a connection string.

Prisma should _never_ use that string if you always provide a custom value via the constructor. If it does anyway, that is cause for creating an bug report here.

**Reply by TheVoid-0** [reply]
Thanks for answering @janpio and sorry for the delay. Providing an URL didn't solve it, it stills randomly tries to connect with the schema url. Me and my team still couldn't reproduce it consistently, but we have  a strong guess this is somewhat related to possible moments where someone makes a request to a client  already  told to `$disconnect`, we've made sure every client is being instantiated with a proper URL, since they are all created via a factory method that throws an error if there's no URL provided. For now, we did as you suggested and provided a mocked url on the schema and took off our connections cache that could be the cause of a client being called after it was meant to be disconnected. It is still soon to say, but so far we've not got another error, I'll try to reproduce the issue more precisely when I find some time.

Thanks again for taking your time to answer

**Reply by janpio** [reply]
If your app ever tried to connect to that mock connection string, you will know something weird is going on. When or if that happens, please open an issue immediately with as much information as possible. That would be really bad and unexpected, so we would really like to know about it to be able to fix it. Thanks.

**Reply by TheVoid-0** [reply]
@janpio I've managed to reproduce it consistently in my application, and as soon as I can isolate from my API logic I will try to create a issue, but just to clarify, my suspicions were correct. The PrismaClient calls the schema.url if someone tries to make a query after the client was `$disconnected`. One could argue, that the correct behavior would be to throw an error indicating the connection termination, or, the client at least should reconnect to the previous host instead of the default one.

to illustrate:

```ts
const connection = TenantConnection.create(url) // this method calls new PrismaClient under the hood

await connection.items.findFirst()

await connection.$disconnect();

await connection.items.findFirst() // <- at this point, the query fails because it will try to connect on the default URL
```

Obviously one should not use the connection after it was closed, but the misleading error strayed us away from what was causing the probem

**Reply by janpio** [reply]
That is an interesting edge case. `$disconnect()` terminates the previous Prisma Client that existed (which was started with a `datasource.url`), and then automatically creates a new one if you send another query. The earlier custom Connection String is not available any more, so it uses the default one. That is most probably not what should happen, so as soon as we have an issue we would try to reproduce this and then see if this needs to (and can) be changed.

**Reply by TheVoid-0** [reply, marked as answer]
@janpio Issue opened! https://github.com/prisma/prisma/issues/12134 Thanks for you attention and help!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
